test/benchmarks/reduce/bench_thrust_reduce.cpp

```diff
@@ -0,0 +1,63 @@
+/* Copyright 2021 Simeon Ehrig
+ *
+ * This file is part of vikunja.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <numeric>
+#include <vector>
+
+#include <catch2/catch.hpp>
+#include <thrust/device_vector.h>
+
+template<typename TData>
+inline void benchmark_reduce(int size)
+{
+    std::vector<TData> hostMemInput(size);
+    for(int i = 0; i < size; ++i)
+    {
+        hostMemInput[i] = static_cast<TData>(i) + static_cast<TData>(1);
+    }
+
+    thrust::device_vector<TData> devMemInput(hostMemInput);
+
+    TData result = thrust::reduce(devMemInput.begin(), devMemInput.end(), static_cast<TData>(0));
+
+    TData expected_result = (static_cast<TData>(size) * (static_cast<TData>(size) + 1)) / static_cast<TData>(2);
+    // verify, that vikunja reduce is working with problem size
+    REQUIRE(expected_result == Approx(result));
+
+    // honeypot to check that the function call in the benchmark block has not been removed by the optimizer
+    result = static_cast<TData>(0);
+
+    BENCHMARK("reduce thrust")
+    {
+        result = thrust::reduce(devMemInput.begin(), devMemInput.end(), static_cast<TData>(0));
+    };
+
+    REQUIRE(expected_result == Approx(result));
+}
+
+TEST_CASE("bechmark reduce - int", "[reduce][thrust][int]")
+{
+    using Data = int;
+    int size = GENERATE(100, 100'000, 1'270'000, 1'600'000);
+    benchmark_reduce<Data>(size);
+}
+
+TEST_CASE("bechmark reduce - float", "[reduce][thrust][float]")
+{
+    using Data = float;
+    int size = GENERATE(100, 100'000, 2'000'000);
+    benchmark_reduce<Data>(size);
+}
+
+TEST_CASE("bechmark reduce - double", "[reduce][thrust][double]")
+{
+    using Data = double;
+    int size = GENERATE(100, 100'000, 1'270'000, 2'000'000);
+    benchmark_reduce<Data>(size);
+}
```

Comment:
I would have preferred a template list test case with an `if constexpr` for the different `GENERATE` expression. But that's just taste I guess.

Reply:
I'm not sure, what you mean.

Reply:
I was suggesting that:
```suggestion
TEMPLATE_TEST_CASE("bechmark reduce", "[reduce][thrust]", int, float, double)
{
    using Data = TestType;
    if constexpr (std::is_same_v<Data, float>)
        benchmark_reduce<TestType>(GENERATE(100, 100'000, 2'000'000));
    else
        benchmark_reduce<TestType>(GENERATE(100, 100'000, 1'270'000, 1'600'000));
}
```
But I now see that each test has different generator values. So I guess you can stick with your version.

Reply:
I understand. But have another reason, why I don't use a `TEMPLATE_TEST_CASE`. If I'm correct, there is no way to generate the test name and tag depending on the template type, which avoids to to run only testcases for specific types. On the other side, maybe this feature is not useful as runtime parameter and makes only sense, if the source code is change, e.g. the problem size was change :thinking:

Reply:
Catch 2 print the type in the test case name. So I overtook your suggestion.